### ISSUE TITLE
Add diagnostic level control to final classed subclass violations

### ISSUE THREAD

**Issue by jgrey4296** [microsoft/pyright#3170]
**Is your feature request related to a problem? Please describe.**
I've been using the `final` class decorator to ensure classes implement all the methods in a collection of abstract protocols.
But, if i want to then have a slight change which just changes a couple of methods, I have to reimplement the entire set of protocols for the new class.
Or I subclass a `final`d class which is guaranteed to implement all the protocols... but then get a complaint that the class is final and can't be subclassed. 

This makes perfect sense, but lacking diagnostic level control I can't turn that complaint down to a warning.
And using `final` in this way seems to be the best way of guarding a class against unimplemented protocol methods, at the class definition, instead of at an instantiation.
Currently pyright doesn't associate a named rule with the error in it's output json. 

**Describe the solution you'd like**
Add a reportSubclassFinalViolation option to enable diagnostic level control.

**Code**
```python
from dataclasses import InitVar, dataclass, field
from typing import final, Protocol
import abc

class MyProto(Protocol):
    @abc.abstractmethod
    def do_something(self) -> str: pass

@final
@dataclass
class Test(MyProto):
    name : str = field()

   def do_something(self):
        return self.name

@dataclass
class More(Test):
    age : int = field()

    def do_something(self):
        return f"{self.name}: {self.age}"

```

**Comment by jgrey4296** [comment]
Somewhat relatedly, and not really worth its own issue, I tried playing with a custom metaclass/decorator to save me needing to decorate every method with `abc.abstractmethod` like below. While it works in the python runtime, pyright complains about the methods which are annotated by the metaclass/decorator:
```python
from dataclasses import InitVar, dataclass, field
from typing import final, Protocol, Any, TypeVar
from types import FunctionType
import abc

class ProtocolMeta(type):
    """
    A Metaclass which creates a protocol,
    and wraps every method defined in abstractmethod
    """

    def __new__(cls, name:str, bases:tuple[type,...], definitions:dict[str,Any]):
        # Check the code of the function is just the "pass" instruction.. I think
        pass_code = b'd\x00S\x00'
        for k,v in definitions.items():
            if (isinstance(v, FunctionType) and v.__code__.co_code == pass_code):
                new_abstract = abc.abstractmethod(v)
                definitions[k] = new_abstract

        if bool(bases):
            bases = (*bases, Protocol)
        else:
            bases = (Protocol,)
        new_protocol = type(name, bases, definitions)

        return new_protocol


T = TypeVar('T')
def my_abstract_protocol(cls : T) -> T:
    """ A class decorator for automatically creating abstractmethods """
    pass_code           = b'd\x00S\x00'
    abstract : set[str] = set()
    for k,v in cls.__dict__.items():
        if (isinstance(v, FunctionType) and v.__code__.co_code == pass_code):
            new_abstract = abc.abstractmethod(v)
            setattr(cls, k, new_abstract)
            abstract.add(k)

    setattr(cls, '__abstractmethods__', frozenset(abstract))
    return cls


class MyProto(metaclass=ProtocolMeta):
    def test(self)     -> str: pass
    def mymethod(self) -> str: pass

@my_abstract_protocol
class MyProto2(Protocol):
    def test(self)     -> str: pass
    def mymethod(self) -> str: pass

@final
@dataclass
class TestMeta(MyProto):
    name : str = field()

    def test(self):
        return self.name

    # def mymethod(self):
    #     return "blah"

@final
@dataclass
class TestDecorator(MyProto2):
    name : str = field()

    def test(self):
        return self.name

    # def mymethod(self):
    #     return "blah"

val  = TestMeta("bob")
val2 = TestDecorator("bob")
print(val)
print(val2)
```

**Comment by erictraut** [comment]
I think the current behavior is correct. You're using `@final` to mean something different than it is intended for. I don't think it make sense to make this a separate diagnostic check.

If you want to use`@final` in this manner, you could use `# type: ignore` to suppress the errors.
